### PR TITLE
✨ Index Fetcher

### PR DESCRIPTION
**Pull request by JordyRo1**
- Adds a new `IndexFetcher` superclass that can create an index given a fetcher and a list of assets/weights.

### CODE REVIEW

File: pragma/publisher/fetchers/index.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    def standardize_decimals(self):
+        decimals = self.asset_weights[0].asset["decimals"]
+
+        for asset_weight in self.asset_weights:
+            asset = asset_weight.asset
+
+            if asset["decimals"] > decimals:
+                exponent = asset["decimals"] - decimals
+
+                for entry in self.spot_entries:
```
[the remark this answers is not in the document]
Comment (JordyRo1) [reply pragma/publisher/fetchers/index.py]: it changes the object itself

File: pragma/publisher/fetchers/index.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        self.asset_weights = asset_weights
+
+    async def fetch(
+        self, session: ClientSession
+    ) -> List[Union[SpotEntry, PublisherFetchError]]:
+        spot_entries = []
+        for asset_weight in self.asset_weights:
+            spot_entry = await self.fetcher._fetch_pair(asset_weight.asset, session)
+            if isinstance(spot_entry, PublisherFetchError):
+                return PublisherFetchError(
```
Comment (EvolveArt) [inline pragma/publisher/fetchers/index.py:49]:
```suggestion
                spot_entry = PublisherFetchError(
```

**Review by EvolveArt — APPROVED** [review]
just the last change and we're good

File: pragma/publisher/fetchers/index.py (hunk 2 above)
Comment (JordyRo1) [reply pragma/publisher/fetchers/index.py:49]: By doing so, we still have to add a check to verify if the spot_entries does not contain a PublisherFetchError.Indeed, we do not want to compute the index value if there is a fetching error in the spot entries array. By returning directly the error, we make sure to handle the error properly and avoid the 'AttributeError: 'PublisherFetchError' object has no attribute 'price'' error instead

Comment (JordyRo1) [reply pragma/publisher/fetchers/index.py:49]: To reproduce the same pattern as for the fetchers, we will need to implement maybe another class with a list of Index to work (analogy to the PRAGMA_ALL_ASSETS used by the fetchers). Currently the IndexFetcher work for a given index_name and List[assets, weights]

**Pull request merged**